contracts/espace/VeShui.sol

```diff
@@ -0,0 +1,341 @@
+            uint256 lockIndex = _dayLocks[i];
+            Lock memory _lock = locks[lockIndex];
+
+            if (!_lock.isHandledLock) {
+                userInfos[_lock.user].totalVeShui += _lock.veShuiAmount;
+                totalVeShui += _lock.veShuiAmount;
+                locks[lockIndex].isHandledLock = true; // storage
+            }
+        }
+        delete dayLocks[day];
+        return totalVeShui;
+    }
+
+    // After handling unlocks, return the totalVeShui of the day, used to calculate dayProfit.
+    // dayProfit = profit / number of days with non-zero veShui
+    // Each day's accCFXPerVeShui += profit / totalVeShui of that day
+    function handleUnlocksOfDay(uint day) private returns (uint256) {
+        uint256[] memory todayUnlocks = dayUnlocks[day];
+        for (uint256 i = 0; i < todayUnlocks.length; i++) {
+            uint256 lockIndex = todayUnlocks[i];
+            Lock memory _lock = locks[lockIndex];
+
+            if (!_lock.isHandledUnlock) {
+                userInfos[_lock.user].totalVeShui -= _lock.veShuiAmount;
+                totalLocked -= _lock.amount;
+            }
+
+            userInfos[_lock.user].unlockedShui += _lock.amount;
+            totalVeShui -= _lock.veShuiAmount;
+
+            delete locks[lockIndex];
+            userInfos[_lock.user].locks.remove(lockIndex);
+        }
+        delete dayUnlocks[day];
+
+        return totalVeShui;
+    }
+
+    function handleProfitOfDay(uint day, uint dayTotalVeshui, uint256 profit) private {
+        accRewardPerVeShui += (profit * 1e18) / dayTotalVeshui;
+
+        address[] memory todayLockUsers = dayLockUsers[day].values();
+        for (uint i = 0; i < todayLockUsers.length; i++) {
+            settleUser(todayLockUsers[i]);
+        }
+        delete dayLockUsers[day];
+
+        address[] memory todayUnlockUsers = dayUnlockUsers[day].values();
+        for (uint i = 0; i < todayUnlockUsers.length; i++) {
+            settleUser(todayUnlockUsers[i]);
+        }
+        delete dayUnlockUsers[day];
+
+        lastSettleDay = day;
+        aprOnLastSeetleDay = (profit * 1e18 * 365) / totalVeShui;
+    }
+
+    function settleUser(address user) private {
+        userInfos[user].reward +=
+            (userInfos[user].totalVeShui * (accRewardPerVeShui - userInfos[user].userRewardPerTokenPaid)) /
+            1e18;
+        userInfos[user].userRewardPerTokenPaid = accRewardPerVeShui;
+    }
+
+    function withdraw() external {
+        uint today = date(block.timestamp);
+        uint256[] memory lockIndexes = userInfos[msg.sender].locks.values();
+
+        for (uint256 i = 0; i < lockIndexes.length; i++) {
+            uint256 lockIndex = lockIndexes[i];
+            if (locks[lockIndex].unlockDay <= today && !locks[lockIndex].isHandledUnlock) {
+                userInfos[msg.sender].unlockedShui += locks[lockIndex].amount;
+                totalLocked -= locks[lockIndex].amount;
+                locks[lockIndex].isHandledUnlock = true;
```

Comment:
should also removed from locks and userInfos.locks, and seems isHandledUnlock is redundant, can be removed